src/sonar_ranger.cpp

```diff
@@ -0,0 +1,216 @@
+
+// Handle sonar range messages and determine distance to obstacle
+
+// Jim Vaughan <jimv@mrjim.com> January 2018
+
+#include <tf2/LinearMath/Transform.h>
+#include <tf2_ros/buffer.h>
+#include <tf2_ros/transform_listener.h>
+#include <sensor_msgs/Range.h>
+#include <visualization_msgs/Marker.h>
+#include "move_basic/sonar_ranger.h"
+
+
+SonarRanger::SonarRanger(): listener(tf_buffer)
+{
+}
+
+void SonarRanger::initialize()
+{
+    ros::NodeHandle nh("");
+    line_pub = ros::Publisher(nh.advertise<visualization_msgs::Marker>("/sonar", 1));
+
+    // TODO: make params
+    std::string topic_prefix = "sonar_";
+
+    for (int i=0; i<16; i++) {
+        std::string topic = str(boost::format{"%1%%2%"} % topic_prefix % i);
+        ROS_INFO("Subscribing to %s", topic.c_str());
+        subscribers.push_back(nh.subscribe("/bus_server/sensor/" + topic, 1, &SonarRanger::callback, this));
+        ids[topic] = i;
+    }
+}
+
+void SonarRanger::callback(const sensor_msgs::Range::ConstPtr &msg)
+{
+    std::string frame = msg->header.frame_id;
+    ROS_INFO("Callback %s", frame.c_str());
+
+    // ignore min values
+    if (msg->range <= msg->min_range || msg->range >= msg->max_range) {
+       return;
+    }
+
+    // create sensor object if this is a new sensor
+    std::map<std::string,SonarSensor>::iterator it = sensors.find(frame);
+    if (it == sensors.end()) {
+        try {
+            geometry_msgs::TransformStamped tfs =
+                tf_buffer.lookupTransform("base_link", frame, ros::Time(0));
```

Comment:
This should be the time of the sensor message for correctness. It shouldn't add any overhead if the transform is static like on our bots, because tf2 would just ignore the time.

Reply:
The current implementation assumes that the sensor don't move.  The origin and vectors are only computed once.

Reply:
Ok, that seems like a reasonable assumption